File: TelerikUIExamples/TelerikUIExamples/DataSourceGettingStarted.m
//
//  DataSourceGettingStarted.m
//  TelerikUIExamples
//
//  Copyright (c) 2015 Telerik. All rights reserved.
//

#import "DataSourceGettingStarted.h"
#import <TelerikUI/TelerikUI.h>

// >> datasource-gettingstarted-full
@interface DataSourceGettingStarted ()

@property (nonatomic, strong) TKDataSource *dataSource;

@end

@implementation DataSourceGettingStarted

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    self.title = @"Getting started";
    
    // >> datasource-getting-started
    self.dataSource = [[TKDataSource alloc] initWithArray:@[ @10, @5, @12, @7, @44 ]];
    // << datasource-getting-started
    
    // >> datasource-data-shaping
    // filter all values less or equal to 5
    [self.dataSource filter:^BOOL(id item) { return [item integerValue] > 5; }];
    
    // sort ascending
    [self.dataSource sort:^NSComparisonResult(id obj1, id obj2) {
        NSInteger a = [obj1 integerValue];
        NSInteger b = [obj2 integerValue];
        if (a<b) { return NSOrderedDescending; }
        else if (a>b) { return NSOrderedAscending; }
        return NSOrderedSame;
    }];
    
    // group odd/even values
    [self.dataSource group:^id(id item) { return @([item integerValue] % 2 == 0); }];
    
    // multiply every value * 10
    [self.dataSource map:^id(id item) { return @([item integerValue] * 10); }];
    
    // find the max value
    NSNumber *maxValue = [self.dataSource reduce:@(0) with:^id(id item, id value) {
        if ([item integerValue] > [value integerValue]) {
            return item;
        }
        return value;
    }];
    // << datasource-data-shaping
    
    NSLog(@"the max value is: %@", maxValue);
    
    // >> datasource-print
    // output everything to the console
    [self.dataSource enumerate:^(id item) {
        if ([item isKindOfClass:[TKDataSourceGroup class]]) {
            TKDataSourceGroup *group = item;
            NSLog(@"group: %@", group.key);
        }
        else {
            NSLog(@"%d", (int)[item integerValue]);
        }
    }];
    // << datasource-print
    
    // >> datasource-tableview
    // bind with a table view
    UITableView *tableView = [[UITableView alloc] initWithFrame:self.view.bounds];
    tableView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    tableView.dataSource = self.dataSource;
    [self.view addSubview:tableView];
    // << datasource-tableview
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end
// << datasource-gettingstarted-full

File: TelerikUIExamplesInSwift/TelerikUIExamplesInSwift/DataSourceGettingStarted.swift
//
//  DataSourceGettingStarted.swift
//  TelerikUIExamplesInSwift
//
//  Copyright (c) 2015 Telerik. All rights reserved.
//

import UIKit

// >> datasource-gettingstarted-full-swift
class DataSourceGettingStarted: TKExamplesExampleViewController {

    var dataSource: TKDataSource?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Do any additional setup after loading the view.
        self.title = "Getting started"
        
        // >> datasource-getting-started-swift
        dataSource = TKDataSource(array: [ 10, 5, 12, 7, 44 ])
        // << datasource-getting-started-swift
        
        // >> datasource-data-shaping-swift
        // filter all values less or equal to 5
        
        dataSource!.filter { (i:AnyObject) -> Bool in
            return (i as! NSNumber).intValue > 5
        }
        
        // sort ascending
        dataSource!.sort {
            let a = $0 as! Int
            let b = $1 as! Int
            if a<b { return NSComparisonResult.OrderedDescending }
            else if a>b { return NSComparisonResult.OrderedAscending }
            return NSComparisonResult.OrderedSame
        }
        
        // group odd/even values
        dataSource!.group { ($0 as! Int) % 2 }
        
        // multiply every value * 10
        dataSource!.map { ($0 as! Int) * 10 }
        
        // find the max value
        let maxValue = dataSource!.reduce(0) {
            if $0 as! Int > $1 as! Int {
                return $0
            }
            return $1
        } as! Int
        // << datasource-data-shaping-swift
        
        print("the max value is: \(maxValue)")
        
        // >> datasource-print-swift
        // output everything to the console
        dataSource!.enumerate {
            if $0.isKindOfClass(TKDataSourceGroup) {
                let group = $0 as! TKDataSourceGroup
                print("group: \(group.key)")
            }
            else {
                print("\($0)")
            }
        }
        // << datasource-print-swift
        
        // >> datasource-tableview-swift
        // bind with a table view
        let tableView = UITableView(frame: self.view.bounds)
        tableView.dataSource = dataSource
        tableView.autoresizingMask = UIViewAutoresizing(rawValue:~UIViewAutoresizing.None.rawValue)
        self.view.addSubview(tableView)
        // << datasource-tableview-swift

    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}
// << datasource-gettingstarted-full-swift
